carma-messenger-core/cpp_message/src/cpp_message.cpp

```diff
@@ -1457,72 +1557,71 @@ namespace cpp_message
+    void Node::encode_day_of_week(const j2735_v2x_msgs::msg::DayOfWeek& msg, DSRC_DayOfWeek_t* output, uint8_t* dow_val)
     {
-        DSRC_DayOfWeek_t* output;
-        output = (DSRC_DayOfWeek_t*) calloc(1, sizeof(DSRC_DayOfWeek_t));
-        
-        uint8_t* dow_val;
-        dow_val = (uint8_t*)calloc(1, sizeof(uint8_t)); // 8 bits are sufficient for bit-wise encoding for 7 days
-        uint8_t tmp_binary;
+        // j2735_v2x_msgs day of week:
+        // Array of [0, 1] for each day, with indices specified as an enum. 1 means that that day is active, 0 means inactive
+        // ex: turn on sunday and monday: [0, 0, 0, 0, 0, 1, 1], dow.SUN -> dow[6] -> 1
+
+        // DSRC day of week:
+        // 7-element bit string with a bit for each day,  means that that day is active, 0 means inactive
+        // ex: turn on sunday and monday: 00000110, dow.DSRC_DayOfWeek_sun -> 1
+
         for (auto i = 0; i < msg.dow.size(); i++)
         {
-            tmp_binary |= msg.dow[i];
-            tmp_binary = tmp_binary << 1;
+            *dow_val |= msg.dow[i];
+            *dow_val = *dow_val << 1;
         }
         output->buf = dow_val;
-        output->size = msg.dow.size();
-
-        return output;
+        output->size = 1; // 1 byte
+        output->bits_unused = 1; // uses the first 7 bits, last is unused
+        return;
     } 
 
-    DailySchedule_t* Node::encode_daily_schedule(const j2735_v2x_msgs::msg::DailySchedule& msg)
+    void Node::encode_daily_schedule(const j2735_v2x_msgs::msg::DailySchedule& msg, DailySchedule_t* output)
     {
-        DailySchedule_t* output;
-        output = (DailySchedule_t*) calloc(1, sizeof(DailySchedule_t));
-        
         output->begin = msg.begin;
         output->duration = msg.duration;
-
-        return output;
+        return;
     } 
 
-    RepeatParams_t* Node::encode_repeat_params (const j2735_v2x_msgs::msg::RepeatParams& msg)
+    void Node::encode_repeat_params (const j2735_v2x_msgs::msg::RepeatParams& msg, RepeatParams_t* output)
     {
-        RepeatParams_t* output;
-        output = (RepeatParams_t*) calloc(1, sizeof(RepeatParams_t));
-
         output->offset = msg.offset;
         output->period = msg.period; 
         output->span = msg.span;   
-
-        return output;
+        return;
     }
 
-    PathNode_t* Node::encode_path_node (const j2735_v2x_msgs::msg::PathNode& msg)
+    void Node::encode_path_node (const j2735_v2x_msgs::msg::PathNode& msg, PathNode_t* output, long* z_temp, long* width_temp)
     {
-        PathNode_t* output;
-        output = (PathNode_t*) calloc(1, sizeof(PathNode_t));
-
         output->x = msg.x;
         output->y = msg.y;
         // optional fields
-        if (msg.z_exists) 
+        if (msg.z_exists)
         {
-            long z_temp = msg.z;
-            output->z = &z_temp;
+            // Because Pathnode.z is a pointer, we have to create a new long variable to store z in (in j2735 it's a short, and segfaults on a pointer cast)
```

Comment:
Open issue?

Reply:
z_temp here is the new long variable. The comment is describing why it's required, but could be explained better